src/cg.jl

```diff
@@ -2,20 +2,29 @@ import Base: iterate
 using Printf
 export cg, cg!, CGIterable, PCGIterable, cg_iterator!, CGStateVariables
 
-mutable struct CGIterable{matT, solT, vecT, numT <: Real}
+# Conjugated dot product
+_dot(x, ::Val{true}) = sum(abs2, x)  # for x::Complex, returns Real
```

Comment:
Maybe better to just use `_norm` / `norm`; that's what was used before and norm is slightly more stable. Also I'm not sure if `sum(abs2, x)` works as efficiently on GPUs, `norm` is definitely optimized.

Reply:
What do you mean by "norm is more stable"?  If you are talking about the possibility of overflow, I think it doesn't matter because the calculated norm is squared again in the algorithm.

I verify that the GPU performance of `sum` does not degrade.
```julia
julia> using LinearAlgebra, CuArrays, BenchmarkTools

julia> v = rand(10^9);

julia> cv = cu(v);

julia> @btime norm($v);
  953.318 ms (0 allocations: 0 bytes)

julia> @btime sum(abs2, $v);
  666.738 ms (0 allocations: 0 bytes)

julia> @btime norm($cv);
  5.764 ms (1 allocation: 16 bytes)

julia> @btime sum(abs2, $cv);
  5.970 ms (365 allocations: 14.00 KiB)
```

Reply:
The main advantage of falling back to `norm` and `dot` in the conjugated case is that they generalize better — if the user has defined some specialized Banach-space type (and corresponding self-adjoint linear operators `A` with overloaded `*`), they should have overloaded `norm` and `dot`, whereas `sum(abs2, x)` might no longer work.

(Even as simple a generalization as an array of arrays will fail with `sum(abs2, x)`, whereas they work with `dot` and `norm`.)

Reply:
> 1.452 s (0 allocations: 0 bytes)  # sum() is significantly faster than norm() on CPU

This is because it dispatches to BLAS which does a stable norm computation. But generally BLAS libs are free to choose how to implement `norm`.

Reply:
> The main advantage of falling back to `norm` and `dot` in the conjugated case is that they generalize better — if the user has defined some specialized Banach-space type (and corresponding self-adjoint linear operators `A` with overloaded `*`), they should have overloaded `norm` and `dot`, whereas `sum(abs2, x)` might no longer work.
> 
> (Even as simple a generalization as an array of arrays will fail with `sum(abs2, x)`, whereas they work with `dot` and `norm`.)

This is a very convincing argument.  Additionally, considering that a user-defined Banach-space type should overload `norm` and `dot` but not the unconjugated dot, it will be nice to be able to define the unconjugated dot using `dot` like
```julia
_dot(x, y, ::UnconjugatedDot) = dot(conj(x), y)
```
but of course this is inefficient as it conjugates `x` twice, not to mention extra allocations.  I wish `dotu` was exported, as you suggested in https://github.com/JuliaLang/julia/issues/22227#issuecomment-306224429.

Users can overload `_dot(x, y, ::UnconjugatedDot)` for the Banach-space type they define, but what would be the best way to minimize such a requirement?

Reply:
You could implement the unconjugated dot as `transpose(x) * y` ... this should have no extra allocations since `transpose` is a "view" type by default?

Reply:
Though `transpose(x) * y` wouldn't work if `x` and `y` are matrices (and `A` is some kind of multlinear operator); maybe it's best to stick with the `sum(zip(x,y))` solution you have now.

Reply:
Can we assume that `getindex()` is always defined for `x` and `y`?  Then how about using `transpose(@view(x[:])) * @view(y[:])`?  This uses allocations, but it seems much faster than `sum(prod, zip(x,y))`:
```
julia> x = rand(100,100); y = rand(100,100);

julia> @btime sum(prod, zip($x,$y));
  12.564 μs (0 allocations: 0 bytes)

julia> @btime transpose(@view($x[:])) * @view($y[:]);
  1.966 μs (4 allocations: 160 bytes)
```

It is nearly on a par with `dot`:
```
julia> @btime dot($x,$y);
  1.839 μs (0 allocations: 0 bytes)
```

Also, when `x` and `y` are `AbstractVectors`, the proposed method does not use allocations:
```
julia> x = rand(1000); y = rand(1000);

julia> @btime transpose(@view($x[:])) * @view($y[:]);
  108.091 ns (0 allocations: 0 bytes)
```

Reply:
You could do `dot(transpose(x'), x)`, maybe?

Reply:
I didn't know that `x'` was nonallocating!  This looked like a great nonallocating implementation, but it turns out that this is slower than the earlier proposed solution using `@view`:
```julia
julia> @btime transpose(@view($x[:])) * @view($y[:]);
  267.794 ns (0 allocations: 0 bytes)

julia> @btime dot(transpose($x'), $y);
  1.547 μs (0 allocations: 0 bytes)
```
Also, neither `x'` nor `transpose(x)` works for `x` whose dimension is greater than 2.

I just pushed an implementation using `@view` for now.